modules/nuxeo-apidoc-webengine/src/main/java/org/nuxeo/apidoc/sitemap/SitemapGenerator.java

```diff
@@ -0,0 +1,98 @@
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * Contributors:
+ *     Anahide Tchertchian
+ */
+package org.nuxeo.apidoc.sitemap;
+
+import static org.nuxeo.apidoc.browse.ApiBrowserConstants.LIST_BUNDLEGROUPS;
+import static org.nuxeo.apidoc.browse.ApiBrowserConstants.LIST_BUNDLES;
+import static org.nuxeo.apidoc.browse.ApiBrowserConstants.LIST_COMPONENTS;
+import static org.nuxeo.apidoc.browse.ApiBrowserConstants.LIST_CONTRIBUTIONS;
+import static org.nuxeo.apidoc.browse.ApiBrowserConstants.LIST_EXTENSIONPOINTS;
+import static org.nuxeo.apidoc.browse.ApiBrowserConstants.LIST_OPERATIONS;
+import static org.nuxeo.apidoc.browse.ApiBrowserConstants.LIST_PACKAGES;
+import static org.nuxeo.apidoc.browse.ApiBrowserConstants.LIST_SERVICES;
+
+import java.io.Writer;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Date;
+import java.util.List;
+import java.util.stream.Collectors;
+
+import javax.xml.bind.JAXBContext;
+import javax.xml.bind.JAXBException;
+import javax.xml.bind.Marshaller;
+
+import org.apache.commons.lang3.StringUtils;
+import org.nuxeo.apidoc.snapshot.DistributionSnapshot;
+import org.nuxeo.apidoc.snapshot.SnapshotManager;
+import org.nuxeo.common.utils.URIUtils;
+import org.nuxeo.ecm.core.api.CoreSession;
+import org.nuxeo.runtime.api.Framework;
+
+/**
+ * @since 20.3.0
+ */
+public class SitemapGenerator {
+
+    protected static final List<String> PATHS = List.of("", LIST_BUNDLEGROUPS, LIST_BUNDLES, LIST_COMPONENTS,
+            LIST_CONTRIBUTIONS, LIST_EXTENSIONPOINTS, LIST_OPERATIONS, LIST_PACKAGES, LIST_SERVICES);
+
+    public static void generateXML(String baseUrl, CoreSession session, Writer writer) throws JAXBException {
+        Sitemap map = getSiteMap(baseUrl, session);
+        JAXBContext context = JAXBContext.newInstance(Sitemap.class);
+        Marshaller mar = context.createMarshaller();
+        mar.setProperty(Marshaller.JAXB_FORMATTED_OUTPUT, Boolean.TRUE);
+        mar.marshal(map, writer);
+    }
+
+    protected static Sitemap getSiteMap(String baseUrl, CoreSession session) {
+        List<SitemapUrl> urls = new ArrayList<>();
+        Framework.getService(SnapshotManager.class)
+                 .getAvailableDistributions(session)
+                 .forEach(s -> urls.addAll(getUrls(baseUrl, s)));
+        return new Sitemap(urls);
+
+    }
+
+    protected static List<SitemapUrl> getUrls(String baseUrl, DistributionSnapshot snapshot) {
+        final Date date = snapshot.getReleaseDate() == null ? snapshot.getCreationDate() : snapshot.getReleaseDate();
```

Comment:
If we're sure that _creationDate_ is never null we can replace the line by:
```suggestion
        Date date = Objects.requireNonNullElseGet(snapshot.getReleaseDate(), snapshot::getCreationDate);
```

If we're not sure _creationDate_ is never null, we could replace by:
```suggestion
        Date date = ObjectUtils.getIfNull(snapshot.getReleaseDate(), snapshot::getCreationDate);
```
There's also a getFirstNonNull which takes a varags Supplier.
Note: You shouldn't need the final keyword, the variable should be explicitly final.

Reply:
creationDate is indeed null in some cases, but ObjectUtils#getIfNull is not present on commons-lang 3.9 (explorer is still aligned on 11.3).
I added the final keyword when i was using a forEach indeed, i'll change the logics

Reply:
Too bad for `ObjectUtils#getIfNull`, we could use the non-functional version `ObjectUtils#firstNonNull`.